include/verilated_saif_c.h

```diff
@@ -0,0 +1,430 @@
+                    bool array, int arraynum);
+};
+
+#ifndef DOXYGEN
+// Declare specialization here as it's used in VerilatedFstC just below
+template <>
+void VerilatedSaif::Super::dump(uint64_t time);
+template <>
+void VerilatedSaif::Super::set_time_unit(const char* unitp);
+template <>
+void VerilatedSaif::Super::set_time_unit(const std::string& unit);
+template <>
+void VerilatedSaif::Super::set_time_resolution(const char* unitp);
+template <>
+void VerilatedSaif::Super::set_time_resolution(const std::string& unit);
+template <>
+void VerilatedSaif::Super::dumpvars(int level, const std::string& hier);
+#endif  // DOXYGEN
+
+//=============================================================================
+// VerilatedSaifBuffer
+
+class VerilatedSaifBuffer VL_NOT_FINAL {
+    // Give the trace file and sub-classes access to the private bits
+    friend VerilatedSaif;
+    friend VerilatedSaif::Super;
+    friend VerilatedSaif::Buffer;
+    friend VerilatedSaif::OffloadBuffer;
+
+    VerilatedSaif& m_owner;  // Trace file owning this buffer. Required by subclasses.
+
+    // CONSTRUCTOR
+    explicit VerilatedSaifBuffer(VerilatedSaif& owner)
+        : m_owner{owner} {}
+    virtual ~VerilatedSaifBuffer() = default;
+
+    //=========================================================================
+    // Implementation of VerilatedTraceBuffer interface
+    // Implementations of duck-typed methods for VerilatedTraceBuffer. These are
+    // called from only one place (the full* methods), so always inline them.
+    VL_ATTR_ALWINLINE void emitEvent(uint32_t code);
+    VL_ATTR_ALWINLINE void emitBit(uint32_t code, CData newval);
+    VL_ATTR_ALWINLINE void emitCData(uint32_t code, CData newval, int bits);
+    VL_ATTR_ALWINLINE void emitSData(uint32_t code, SData newval, int bits);
+    VL_ATTR_ALWINLINE void emitIData(uint32_t code, IData newval, int bits);
+    VL_ATTR_ALWINLINE void emitQData(uint32_t code, QData newval, int bits);
+    VL_ATTR_ALWINLINE void emitWData(uint32_t code, const WData* newvalp, int bits);
+    VL_ATTR_ALWINLINE void emitDouble(uint32_t code, double newval);
+};
+
+//=============================================================================
+// VerilatedFile
+/// Class representing a file to write to. These virtual methods can be
+/// overrode for e.g. socket I/O.
+
+class VerilatedSaifFile VL_NOT_FINAL {
+private:
+    int m_fd = 0;  // File descriptor we're writing to
+public:
+    // METHODS
+    /// Construct a (as yet) closed file
+    VerilatedSaifFile() = default;
+    /// Close and destruct
+    virtual ~VerilatedSaifFile() = default;
+    /// Open a file with given filename
+    virtual bool open(const std::string& name) VL_MT_UNSAFE;
+    /// Close object's file
+    virtual void close() VL_MT_UNSAFE;
+    /// Write data to file (if it is open)
+    virtual ssize_t write(const char* bufp, ssize_t len) VL_MT_UNSAFE;
+};
```

Comment:
Do we actually need this class? I can see the copy-pasta comment about socket I/O that came from the VCD implementation, but given that SAIF is not a streamable file format, is this useful? FST doesn't have it as there is no hook in the FST API to hand off write data. I'm mostly concerned with whether this is unnecessary complexity. @wsnyder to comment please.